kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java

```diff
@@ -0,0 +1,164 @@
+
+    // create one replica for A, two for B
+    replicaMetadataStore.createSync(
+        new ReplicaMetadata(UUID.randomUUID().toString(), snapshotA.snapshotId));
+    replicaMetadataStore.createSync(
+        new ReplicaMetadata(UUID.randomUUID().toString(), snapshotB.snapshotId));
+    replicaMetadataStore.createSync(
+        new ReplicaMetadata(UUID.randomUUID().toString(), snapshotB.snapshotId));
+
+    ReplicaCreatorService replicaCreatorService =
+        new ReplicaCreatorService(metadataStoreService, REPLICAS_PER_SNAPSHOT_TEST, meterRegistry);
+    replicaCreatorService.startAsync();
+    replicaCreatorService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    assertThat(MetricsUtil.getCount(ReplicaCreatorService.REPLICAS_CREATED, meterRegistry))
+        .isEqualTo(0);
+
+    snapshotMetadataStore.close();
+    replicaMetadataStore.close();
+
+    metadataStoreService.stopAsync();
+    metadataStoreService.awaitTerminated(DEFAULT_START_STOP_DURATION);
+  }
+
+  @Test
+  public void shouldCreateReplicasIfNoneExist() throws Exception {
+    KaldbConfigs.ZookeeperConfig zkConfig =
+        KaldbConfigs.ZookeeperConfig.newBuilder()
+            .setZkConnectString(testingServer.getConnectString())
+            .setZkPathPrefix("shouldCreateReplicasIfNoneExist")
+            .setZkSessionTimeoutMs(1000)
+            .setZkConnectionTimeoutMs(1000)
+            .setSleepBetweenRetriesMs(1000)
+            .build();
+
+    MetadataStoreService metadataStoreService = new MetadataStoreService(meterRegistry, zkConfig);
+    metadataStoreService.startAsync();
+    metadataStoreService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    SnapshotMetadataStore snapshotMetadataStore =
+        new SnapshotMetadataStore(metadataStoreService.getMetadataStore(), true);
+
+    SnapshotMetadata snapshotA =
+        new SnapshotMetadata(
+            "a", "a", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "a");
+    SnapshotMetadata snapshotB =
+        new SnapshotMetadata(
+            "b", "b", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "b");
+
+    snapshotMetadataStore.createSync(snapshotA);
+    snapshotMetadataStore.createSync(snapshotB);
+
+    ReplicaCreatorService replicaCreatorService =
+        new ReplicaCreatorService(metadataStoreService, REPLICAS_PER_SNAPSHOT_TEST, meterRegistry);
+    replicaCreatorService.startAsync();
+    replicaCreatorService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    await()
+        .until(
+            () ->
+                MetricsUtil.getCount(ReplicaCreatorService.REPLICAS_CREATED, meterRegistry)
+                    == REPLICAS_PER_SNAPSHOT_TEST * 2);
+
+    ReplicaMetadataStore replicaMetadataStore =
+        new ReplicaMetadataStore(metadataStoreService.getMetadataStore(), true);
+    assertThat(replicaMetadataStore.getCached().size()).isEqualTo(4);
+    assertThat(
+            (int)
+                replicaMetadataStore
+                    .getCached()
```

Comment:
In unit tests prefer listSync calls to check metadata instead of cached calls which may be lagging due to timing issues.

I think it would be a good idea to add a check to make sure that the existing data before the assignment is still the same after the assignment.

Reply:
As the underlying implementation is written against the cached view, when the test runs it should be using the same underlying data structure. 

I can add an an extra `await()` here however. Technically in line 132 we've already guaranteed that the service has bootstrapped and initialized all the data structures, but it would be possible under heavy load that the cache view hasn't been updated yet.

Reply:
Per our discussion I went ahead and added an extra check to `listSync`, as well as the `await()` on the `getCached` call.